Migrate `login` to version catalogs

wzieba opened the pull request:
### Description

This PR migrates `login` module to use local version catalog and unifies setup of per-project repositories for easier management. 

### Testing

This PR introduces build-related changes - a green light from CI and empty list of dependency diff should be just fine.

wzieba committed:
login: use WordPress utils and lint checks from version catalog

wzieba committed:
login: use appcompat from version catalog

wzieba committed:
login: use constraintlayout from version catalog

wzieba committed:
login: use androidx core from version catalog

wzieba committed:
login: use play service auth from version catalog

wzieba committed:
login: use glide from version catalog

wzieba committed:
login: migrate and use androidx credentials to version catalog

dangermattic commented:
<!--
  0 Errors
  0 Warnings
  1 Message: This PR is still a Draft: some...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

wzieba committed:
Merge branch 'post_fluxc_merge_build_improvements' into migrate_login_to_version_catalogs

wzieba committed:
login: use dagger from version catalog

wzieba committed:
login: use junit from version catalog

wzieba committed:
login: use mockito from version catalog

wzieba committed:
login: use androidx core testing from version catalog

wzieba committed:
login: use robolectric from version catalog

wzieba committed:
login: use assertj from version catalog

wzieba committed:
login: use google material from version catalog

wzieba committed:
login: use migrate gravatar to use from version catalog

wzieba committed:
login: remove unused `ext`

wzieba committed:
Unify project's repositories

It's easier to manage a unified list of repositories, instead of having to declare them in each project

wzieba committed:
Cleanup jcenter-mirror inclusions

wzieba committed:
Ignore false-positive `CredentialDependency` lint warning

wzieba committed:
Bring back Dependabot (which I unsuccessfully tried to limit to version catalog only in 357aba1fa3227520fe9f6c79cb364cd27f841314)

wzieba committed:
Bump Dagger to 2.52

The project was using Dagger 2.52 already anyway, as 2.52 was requested by `login` project

wzieba commented on libs/login/build.gradle:
```diff
@@ -36,68 +23,40 @@ android {
     // Dagger
-    implementation "com.google.dagger:dagger:$daggerVersion"
-    kapt "com.google.dagger:dagger-compiler:$daggerVersion"
-    compileOnly "org.glassfish:javax.annotation:$javaxAnnotationVersion"
```
This dependency was removed intentionally as it doesn't seem to be used anywhere. The compilation was successful without it.

ParaskP7 commented:
FYI: @wzieba , in the previous #21353 PR I forgot to mention about this `androidx-paging-runtime` version and the recommendation to use `androidx-paging` instead. The `runtime` suffix can be removed and only be utilized on the dependency itself, which could be `runtime` related, `ktx` related, or any other `androidx-paging` related dependency. However, the same `androidx-paging` version should apply to all these dependencies, this version shouldn't be specific to `runtime`. Hope that makes sense. 🙏 

PS: As an example, check this `androidx-room` version you also migrated in that PR.

ParaskP7 commented on .github/dependabot.yml:
```diff
@@ -2,7 +2,7 @@ version: 2
   - package-ecosystem: "gradle"
     open-pull-requests-limit: 6
-    directory: "gradle/"
+    directory: "/"
```
Praise (❤️): Thanks for reverting this change! 💯

ParaskP7 commented on config/lint/lint.xml:
```diff
@@ -4,6 +4,7 @@
     <issue id="MissingTranslation" severity="ignore" />
     <issue id="ExtraTranslation" severity="ignore" />
     <issue id="PluralsCandidate" severity="ignore" /> <!-- GlotPress doesn't support plurals -->
+    <issue id="CredentialDependency" severity="ignore" /> <!-- False positive, fixed in AGP 8.6.0: https://issuetracker.google.com/issues/339841694 -->
```
Praise (❤️): 👍 on that and I guess that similarly we can ignore [this](https://github.com/wordpress-mobile/WordPress-Android/security/code-scanning/11376) core scanning security alert.

ParaskP7 commented on gradle/libs.versions.toml:
```diff
@@ -55,7 +56,7 @@ facebook-shimmer = '0.5.0'
 fladle = '0.17.5'
 google-autoservice = '1.0-rc4'
-google-dagger = '2.50'
+google-dagger = '2.52'
```
Praise (❤️): Awesome! 🥇

ParaskP7 commented on gradle/libs.versions.toml:
```diff
@@ -141,6 +142,8 @@ androidx-constraintlayout-compose = { group = "androidx.constraintlayout", name
 androidx-constraintlayout-main = { group = "androidx.constraintlayout", name = "constraintlayout", version.ref = "androidx-constraintlayout-main" }
 androidx-core-ktx = { group = "androidx.core", name = "core-ktx", version.ref = "androidx-core" }
 androidx-core-main = { group = "androidx.core", name = "core", version.ref = "androidx-core" }
+androidx-credentials = { module = "androidx.credentials:credentials", version.ref = "androidx-credentials" }
```
Minor (🔍): Consider adding the `main` suffix to this dependency to differential it from being both a dependency and a dependency group too, just like it was done for all other dependencies.

wzieba replied:
Sure, I'll follow this convention, but I'm unsure if I see a value in adding this suffix - what's the problem of having simply `libs.androidx.core` instead of `libs.androidx.core.main`?

ParaskP7 replied:
Great question @wzieba ! 💯 

I tried to summarize that in those version catalogs migration PRs of mine ([example](https://github.com/wordpress-mobile/WordPress-Android/pull/21262)):

> Where needed the word main is added in its suffix to make sure that a dependency cannot be a parent of another dependency and a dependency itself. This helps in various corner cases.

I didn't want to complicate things and try to give such corner cases, but there are a few. For example, consider this line [here](https://github.com/wordpress-mobile/WordPress-Android/blob/trunk/WordPress/build.gradle#L525) (`libs.androidx.test.espresso.core.get().version`). If you go to the version catalog itself and remove the `-core` suffix, then try to sync, you will immediately notice the following build error:

```
No signature of method: org.gradle.accessors.dm.LibrariesForLibs$AndroidxTestEspressoLibraryAccessors.get() is applicable for argument types: () values: []
```

So, although defining `androidx-test-espresso` to point to `espresso-core` and `androidx-test-espresso-contrib` to point to `espresso-contrib` is perfectly fine when those are using as dependencies, when you start trying to do something more, then you get into trouble.

PS: Think about this `-core` being pretty similar to `-main`, it is just that sometimes you don't have this `core` naming convention on a dependency signature. Initially, I started naming everything with the `-core` suffix, but then I realized that some dependencies have both a main dependency and a core dependency (see [androidx-lifecycle-livedata](https://github.com/wordpress-mobile/WordPress-Android/blob/trunk/gradle/libs.versions.toml#L148C1-L149)), plus no dependency have the `main` naming convention, so it was free for me to pick-up  and thus I went with this naming convention (instead of `core`). Also see this [androidx-core](https://github.com/wordpress-mobile/WordPress-Android/blob/trunk/gradle/libs.versions.toml#L142C1-L143) and imagine having to name this as `androidx-core-core` (😅).

Does this help? 🙏

wzieba replied:
> If you go to the version catalog itself and remove the -core suffix

Sorry, I'm not following - why would I want to remove `-core` suffix, if it's in the name of dependency?

What I mean is that I don't see value in adding artificial suffixes, meaning suffixes that are not part of a dependency name.

`androidx-test-espresso-core` is fine to me because module is `androidx.test.espresso:espresso-core`. 
`androidx-credentials-main` is not good idea IMO, as module is `androidx.credentials:credentials` so "main" came out of nowhere.

wzieba replied:
Ah ok, I get it: `libs.androidx.test.espresso` points to the group accessor, instead of the specific dependency.

ParaskP7 replied:
Yea, apologies for trying to explain it in such a complicated way, I could have definitely done a better job, I was just trying to find a example that makes sense and you can quickly see the build erroring.... 🫣

ParaskP7 commented on settings.gradle:
```diff
@@ -16,6 +16,50 @@ plugins {
     id "com.gradle.develocity" version "3.18.1"
 }
 
+dependencyResolutionManagement {
```
Praise (❤️): Great change! 🥇 

PS: Same for removing the unneeded `includeVersion` from `https://a8c-libs.s3.amazonaws.com/android/jcenter-mirror`. 💯

wzieba committed:
Rename androidx-paging-runtime to androidx-paging

ParaskP7 reviewed (approved):
👋 @wzieba !

I have reviewed and quickly smoke tested this PR too, everything works as expected, excellent work! 🌟

FYI: This extra standalone comment: https://github.com/wordpress-mobile/WordPress-Android/pull/21367#issuecomment-2444358946

wzieba committed:
Add `.main` suffix to credentials dependency

wzieba committed:
Merge branch 'trunk' into migrate_login_to_version_catalogs

wzieba commented:
Thanks for both reviews @ParaskP7 ! 🙌
